rules/NoConfigResolverParametersInConstructorRule.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\PHPStan\Rules;
+
+use Ibexa\Contracts\Core\SiteAccess\ConfigResolverInterface;
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\Type\ObjectType;
+
+/**
+ * @implements \PHPStan\Rules\Rule<\PhpParser\Node\Expr\MethodCall>
+ */
+final class NoConfigResolverParametersInConstructorRule implements Rule
+{
+    public function getNodeType(): string
+    {
+        return Node\Expr\MethodCall::class;
+    }
+
+    /**
+     * @throws \PHPStan\ShouldNotHappenException
+     */
+    public function processNode(Node $node, Scope $scope): array
+    {
+        if (!$node->name instanceof Node\Identifier) {
+            return [];
+        }
+
+        $function = $scope->getFunction();
+        if ($function !== null && $function->getName() !== '__construct') {
+            return [];
+        }
+
+        /** @var \PhpParser\Node\Identifier $nodeName */
+        $nodeName = $node->name;
+        $methodName = $nodeName->name;
+
+        if (
+            $methodName !== 'getParameter'
+            && $methodName !== 'hasParameter'
+            && !isset($node->getArgs()[0])
+        ) {
+            return [];
+        }
+
+        $type = $scope->getType($node->var);
+        $configResolverInterfaceType = new ObjectType(ConfigResolverInterface::class);
+        if (!$configResolverInterfaceType->isSuperTypeOf($type)->yes()) {
+            return [];
+        }
+
+        return [
+            RuleErrorBuilder
+                ::message('Referring to ConfigResolver parameters in constructor is not allowed due to potential scope change.')
+                ->identifier('Ibexa.NoConfigResolverParametersInConstructor')
+                ->nonIgnorable()
```

Comment:
Personally given that this is an issue that does not immediately break the app, it should be ignorable. Within team we will make sure that it's not ignored anyway, while anyone using those rules can choose to ignore it if they wish - just like any other PHPStan error.

Reply:
I beg to disagree - moving it outside constructor isn't a big deal and in fact should be spread as not only a good but necessary practice. We can discuss it broader with whoever cares though.

Reply:
Besides - we block such approach on code review anyway, so... 😉

Reply:
As this is a new rule, it will not exist in any baselines yet.

All I'm saying is that this option was not meant as far as I understand it to enforce anything, and it's almost like `!important` in a CSS. It's available, but it should not be used.

Reply:
Well, I am still not convinced as I treat it as an error in the context of being Ibexa DXP user so I suggest we decide within the team.

Reply:
Wise man once said: Here's a rope. You can do what you want with it :sweat_smile: 

I prefer to give the developer an option, than not giving it. What they do with the rule is up to them. Forcing someone to fix their code immediately is a thing most tools (PHPStan, `git`, and others) try to avoid.

But regardless, I'm okay with the current state. For the moment it's a detail that will probably never have any effect.

Reply:
After internal discussions we will go for `nonIgnorable` and introduce a way to disable specific rules later on.